File: Software/C#/FreETarget.NET/FreETarget.NET.Data/Models/Svg/SvgImage.cs
using System.Xml;

namespace FreeTargetWeb01.Models.Svg
{
    public class SvgImage
    {
        public int Width { get; set; }
        public int Height { get; set; }

        public string BackgroundColor { get; set; } = "#FFE4B5";
        public string TextColor { get; set; } = "black";

        public String TextFont { get; set; } = "Arial";
        public List<SvgElement> Elements { get; set; } = new List<SvgElement>();

        public void AddCircle(int x, int y, int r, string? fill = null, string? stroke = null, int? strokeWidth = null)
        {
            SvgElement svgElement = new SvgElement
            {
                ElementType = SvgElementType.Circle,
                X = x,
                Y = y,
                R = r,
                Fill = fill,
                Stroke = stroke,
                StrokeWidth = strokeWidth
            };
            Elements.Add(svgElement);
        }

        public void AddRectangle(int x, int y, int width, int height, string? fill = null)
        {
            SvgElement svgElement = new SvgElement
            {
                ElementType = SvgElementType.Rectangle,
                X = x,
                Y = y,
                Width = width,
                Height = height
            };
            if (fill != null)
            {
                svgElement.Fill = fill;
            }
            Elements.Add(svgElement);
        }

        public void AddText(int x, int y, string text, int fontSize, string? fill = null, string? fontWeight = null )
        {
            SvgElement svgElement = new SvgElement
            {
                ElementType = SvgElementType.Text,
                X = x,
                Y = y,
                Text = text,
                FontSize = fontSize,
                FontWeight = fontWeight
            };
            if (fill != null)
            {
                svgElement.Fill = fill;
            }
            Elements.Add(svgElement);
        }

        public override string ToString()
        {
            string resultat = "";
            using (var sw = new StringWriter())
            {
                using (var xmlWriter = XmlWriter.Create(sw))
                {
                    const string svgNs = "http://www.w3.org/2000/svg";
                    xmlWriter.WriteStartDocument();
                    xmlWriter.WriteStartElement("svg", svgNs);
                    //xmlWriter.WriteAttributeString("width", Width.ToString());
                    //xmlWriter.WriteAttributeString("height", Height.ToString());
                    xmlWriter.WriteAttributeString("viewBox", $"0 0 {Width} {Height}");

                    if (TextFont != null)
                    {
                        xmlWriter.WriteStartElement("style");
                        xmlWriter.WriteString($"text {{font-family: {TextFont} }}");
                        xmlWriter.WriteEndElement();
                    }

                    foreach (var element in Elements)
                    {
                        switch (element.ElementType)
                        {
                            case SvgElementType.Circle:
                                xmlWriter.WriteStartElement("circle");
                                xmlWriter.WriteAttributeString("cx", element.X.ToString());
                                xmlWriter.WriteAttributeString("cy", element.Y.ToString());
                                xmlWriter.WriteAttributeString("r", element.R.ToString());
                                if (element.Fill != null)
                                {
                                    xmlWriter.WriteAttributeString("fill", element.Fill);
                                }
                                if (element.Stroke != null)
                                {
                                    xmlWriter.WriteAttributeString("stroke", element.Stroke);
                                }
                                if (element.StrokeWidth != null)
                                {
                                    xmlWriter.WriteAttributeString("stroke-width", element.StrokeWidth.ToString());
                                }
                                xmlWriter.WriteEndElement();
                                break;
                            case SvgElementType.Rectangle:
                                xmlWriter.WriteStartElement("rect");
                                xmlWriter.WriteAttributeString("x", element.X.ToString());
                                xmlWriter.WriteAttributeString("y", element.Y.ToString());
                                xmlWriter.WriteAttributeString("width", element.Width.ToString());
                                xmlWriter.WriteAttributeString("height", element.Height.ToString());
                                if (element.Fill != null)
                                {
                                    xmlWriter.WriteAttributeString("fill", element.Fill);
                                }
                                xmlWriter.WriteEndElement();
                                break;
                            case SvgElementType.Text:
                                xmlWriter.WriteStartElement("text");
                                xmlWriter.WriteAttributeString("x", element.X.ToString());
                                xmlWriter.WriteAttributeString("y", element.Y.ToString());
                                xmlWriter.WriteAttributeString("font-size", element.FontSize.ToString());
                                xmlWriter.WriteAttributeString("dominant-baseline", "middle");
                                xmlWriter.WriteAttributeString("text-anchor", "middle");

                                if (element.Fill != null)
                                {
                                    xmlWriter.WriteAttributeString("fill", element.Fill);
                                }

                                if (element.FontWeight != null)
                                {
                                    xmlWriter.WriteAttributeString("font-weight", element.FontWeight);
                                }

                                xmlWriter.WriteString(element.Text);
                                xmlWriter.WriteEndElement();
                                break;
                        }
                    }
                    xmlWriter.WriteEndElement();
                    xmlWriter.WriteEndDocument();

                }
                resultat = sw.ToString();
                resultat = resultat.Substring(resultat.IndexOf("?>") + 2);

            }
            return resultat;
        }
    }
}

File: Software/BrowserClientPOC/src/svg/svgImage.ts
import { SvgElement } from "./svgElement";
import { SvgElementType } from "./svgElementType";

export class SvgImage {
    width: number;
    height: number;
    textFont: string | "Arial";
    elements: SvgElement[];

    constructor(w: number, h: number) {
        this.width = w;
        this.height = h;

        this.elements = [];
    }

    addCircle(x: number, y: number, r: number, fill?: string | null, stroke?: string | null, strokeWidth?: number | null) {
        let element = new SvgElement();
        element.elementType = SvgElementType.Circle;
        element.x = x;
        element.y = y;
        element.r = r;
        if (fill) {
            element.fill = fill;
        }

        if (stroke) {
            element.stroke = stroke;
        }
        if (strokeWidth) {
            element.strokeWidth = strokeWidth;
        }
        this.elements.push(element);
    }

    addLine(x: number, y: number, x2: number, y2: number, fill?: string | null, stroke?: string | null, strokeWidth?: number | null) {
        let element = new SvgElement();
        element.elementType = SvgElementType.Line;
        element.x = x;
        element.y = y;
        element.x2 = x2;
        element.y2 = y2;
        if (fill) {
            element.fill = fill;
        }
        if (stroke) {
            element.stroke = stroke;
        }
        if (strokeWidth) {
            element.strokeWidth = strokeWidth;
        }
        this.elements.push(element);
    }

    addRectangle(x: number, y: number, width: number, height: number, fill?: string | null) {
        let element = new SvgElement();
        element.elementType = SvgElementType.Rectangle;
        element.x = x;
        element.y = y;
        element.width = width;
        element.height = height;
        if (fill) {
            element.fill = fill;
        }
        this.elements.push(element);
    }

    addText(x: number, y: number, text: string, fontSize: number, fill?: string | null, fontWeight?: string | null) {

        let element = new SvgElement();
        element.elementType = SvgElementType.Text;
        element.x = x;
        element.y = y;
        element.text = text;
        element.fontSize = fontSize;
        if (fontWeight) {
            element.fontWeight = fontWeight;
        }
        if (fill) {
            element.fill = fill;
        }
        this.elements.push(element);
    }

    toString(): string {

        let svg = `<svg viewBox="0 0 ${this.width} ${this.height}" xmlns="http://www.w3.org/2000/svg">`;

        if (this.textFont) {
            svg += `<style>
            text {
                font-family: ${this.textFont};
            }
            </style>`;
        }

        for (let element of this.elements) {
            switch (element.elementType) {
                case SvgElementType.Rectangle:
                    svg += `<rect x="${element.x}" y="${element.y}" width="${element.width}" height="${element.height}"`;
                    if (element.fill) {
                        svg += ` fill="${element.fill}"`;
                    }
                    svg += ` />`;
                    break;
                case SvgElementType.Circle:
                    svg += `<circle cx="${element.x}" cy="${element.y}" r="${element.r}"`;
                    if (element.fill) {
                        svg += ` fill="${element.fill}"`;
                    }
                    if (element.stroke) {
                        svg += ` stroke="${element.stroke}"`;
                    }
                    if (element.strokeWidth) {
                        svg += ` stroke-width="${element.strokeWidth}"`;
                    }
                    svg += ` />`;
                    break;
                case SvgElementType.Text:
                    svg += `<text x="${element.x}" y="${element.y}" font-family="${this.textFont}" `;
                    if (element.fill) {
                        svg += ` fill="${element.fill}"`;
                    }
                    if (element.fontWeight) {
                        svg += ` font-weight="${element.fontWeight}"`;
                    }
                    if (element.fontSize) {
                        svg += ` font-size="${element.fontSize}"`;
                    }
                    svg += `>${element.text}</text> />`;

                    break;
                case SvgElementType.Line:
                    svg += `<line x1="${element.x}" y1="${element.y}" x2="${element.x2}" y2="${element.y2}"`;
                    if (element.fill) {
                        svg += ` fill="${element.fill}"`;
                    }
                    if (element.stroke) {
                        svg += ` stroke="${element.stroke}"`;
                    }
                    if (element.strokeWidth) {
                        svg += ` stroke-width="${element.strokeWidth}"`;
                    }
                    svg += ` />`;
            }
        }

        svg += `</svg>`;
        return svg;
    }


}